src/concatenate.js

```diff
@@ -0,0 +1,282 @@
+        return;
+      }
+
+      allPlaylists.forEach((playlist) => {
+        playlist.segments = resolvedPlaylistsMap[playlist.resolvedUri].segments;
+      });
+
+      const combinedVideoPlaylist = combinePlaylists({ playlists: videoPlaylists });
+      const combinedAudioPlaylist = audioPlaylists.length ? combinePlaylists({
+        playlists: audioPlaylists,
+        uriSuffix: '-audio'
+      }) : null;
+      const manifestObject = constructMasterManifest({
+        videoPlaylist: combinedVideoPlaylist,
+        audioPlaylist: combinedAudioPlaylist
+      });
+      const concatenationResult = { manifestObject };
+
+      callback(null, concatenationResult);
+    }
+  });
+};
+
+/**
+ * @typedef {Object} ManifestConfig
+ * @property {string} url
+ *           URL to the manifest
+ * @property {string} mimeType
+ *           Mime type of the manifest
+ *           @see {@link https://www.w3.org/TR/html51/semantics-embedded-content.html#mime-types|Mime Types}
+ */
+
+/**
+ * Returns an error message if there's an issue with the user provided manifest objects,
+ * or returns null if no error (at the top level).
+ *
+ * @param {ManifestConfig[]} manifests
+ *        Array of ManifestConfig objects
+ * @return {string|null}
+ *         Error message or null if no error
+ */
+export const getProvidedManifestsError = (manifests) => {
+  if (!manifests || !manifests.length) {
+    return 'No sources provided';
+  }
+
+  for (let i = 0; i < manifests.length; i++) {
+    // The requirement for every manifest needing a URL may be reconsidered in the future
+    // to accept pre-parsed manifest objects.
+    if (!manifests[i].url) {
+      return 'All manifests must include a URL';
+    }
+
+    if (!manifests[i].mimeType) {
+      return 'All manifests must include a mime type';
+    }
+  }
+
+  return null;
+};
+
+/**
+ * @typedef {Object} ConcatenationResult
+ * @property {Object} manifestObject
+ *           Concatenated manifest object
+ */
+
+/**
+ * Calls back with a single rendition VHS formatted master playlist object given a list of
+ * URLs and their mime types as well as a target vertical resolution.
+ *
+ * As of now, only DASH and HLS are supported.
+ *
+ * This function will select the closest rendition (absolute value difference) to the
+ * target vertical resolution. If resolution information is not available as part of the
+ * manifest, then it will fall back to the INITIAL_BANDWIDTH config value from VHS.
+ *
+ * @param {Object} config
+ *        Object containing arguments
+ * @param {ManifestConfig[]} config.manifests
+ *        Array of manifest config objects
+ * @param {number} config.targetVerticalResolution
+ *        The vertical resolution to search for among playlists within each manifest
+ * @param {function(Object, ConcatenationResult)} config.callback
+ *        Callback function with error and result object parameters
+ */
+export const concatenateVideos = ({ manifests, targetVerticalResolution, callback }) => {
```

Comment:
what's the browser support like for this plugin? Promises would probably make the code a lot cleaner compared to callbacks.